### PR TITLE
Created WALGoHeadHashdex

### PR DESCRIPTION
**Pull request by cherep58**

### CODE REVIEW

File: pp-pkg/storage/adapter.go [hunk 1]
```diff
@@ -87,28 +87,28 @@ func (ar *Adapter) AppendHashdex(
...
-	var floatsAppended float64
 	defer func(start time.Time) {
 		ar.appendDuration.Observe(float64(time.Since(start).Microseconds()))
-		ar.samplesAppended.Add(floatsAppended)
+		ar.samplesAppended.Add(float64(stats.SamplesAdded))
 	}(time.Now())
 
-	return ar.proxy.With(ctx, func(h *pp_storage.Head) error {
-		stats, err := appender.New(h, services.CFViaRange).Append(
+	_ = ar.proxy.With(ctx, func(h *pp_storage.Head) error {
```
Comment (vporoshok) [inline pp-pkg/storage/adapter.go]: The error from `proxy.With` is now silently discarded. Previously `AppendHashdex` returned it directly via `return ar.proxy.With(...)`, so callers (`refill_processor`, `stream_processor`) could observe proxy-level errors (e.g. context cancellation, head unavailability).

Also note the inconsistency: `AppendSnappyProtobuf` (line 157) still uses `return ar.proxy.With(...)` and propagates proxy errors, while `AppendHashdex` does not.

Please either:
- Propagate the proxy error here as well (and update the pattern for consistency), or
- Verify that `proxy.With` cannot return its own error apart from the callback, and add a comment explaining the pattern.

File: pp/entrypoint/wal_hashdex.cpp [hunk 2]
```diff
@@ -193,3 +194,22 @@ extern "C" void prompp_wal_open_metrics_scraper_hashdex_parse(void* args, void*
...
+
+extern "C" void prompp_wal_go_head_hashdex_presharding(void* args) {
+  struct Arguments {
+    HashdexVariant* hashdex;
+    entrypoint::head::LssVariantPtr lss;
+    entrypoint::head::DataStoragePtr data_storage;
+  };
+
+  const auto in = static_cast<Arguments*>(args);
+  std::get<GoHeadHashdex>(*in->hashdex).presharding(&std::get<entrypoint::head::QueryableEncodingBimap>(*in->lss), in->data_storage.get());
```
Comment (vporoshok) [inline pp/entrypoint/wal_hashdex.cpp:214]: All other presharding functions (`prompp_wal_protobuf_hashdex_snappy_presharding`, `prompp_wal_go_model_hashdex_presharding`) wrap the C++ logic in `try/catch` and return exceptions as a Go byte slice. This function has no exception handling.

If any exception is thrown (e.g. from `OutdatedChunkMerger::merge()` or hash computation), it will propagate across the CGo boundary — this is undefined behavior and will likely crash the process.

Please add a `try/catch` block with error return, following the same pattern as the other presharding functions.

File: pp/go/cppbridge/wal_hashdex.go [hunk 3]
```diff
@@ -411,3 +411,41 @@ func (*WALOpenMetricsScraperHashdex) Replica() string {
...
+
+func NewGoHeadHashdex(lss *LabelSetStorage, dataStorage *DataStorage) *WALGoHeadHashdex {
+	hashdex := &WALGoHeadHashdex{
+		hashdex:     walGoHeadHashdexCtor(),
+		lss:         lss,
+		dataStorage: dataStorage,
+	}
+	runtime.SetFinalizer(hashdex, func(hashdex *WALGoHeadHashdex) {
+		walHashdexDtor(hashdex.hashdex)
+	})
```
Comment (vporoshok) [inline pp/go/cppbridge/wal_hashdex.go:447]: nit: For consistency with `WALBasicDecoderHashdex` (which calls `runtime.KeepAlive(h.metadata)` in the finalizer), consider adding `runtime.KeepAlive(hashdex.lss)` and `runtime.KeepAlive(hashdex.dataStorage)` here.

Technically safe without them (Go finalizer ordering guarantees these won't be collected before this finalizer runs), but it makes the intent explicit and is a defensive safeguard.

File: pp-pkg/storage/batch_storage.go [hunk 4]
```diff
@@ -20,9 +19,8 @@ type BatchStorage struct {
 	hashdexLimits   cppbridge.WALHashdexLimits
 	transactionHead *pp_storage.TransactionHead
 	state           *cppbridge.StateV2
-	// TODO tmp
-	batch   *timeSeriesBatch
-	adapter *Adapter
+	adapter         *Adapter
+	samplesAdded    uint32
```
Comment (vporoshok) [inline pp-pkg/storage/batch_storage.go:23]: nit: The `timeSeriesBatch` type (in `appender.go`) is no longer used by `BatchStorage`. The doc comment on line 15 still says "aggregates the `[model.TimeSeries]` batch" — consider updating it to reflect the new behavior.

File: pp/primitives/label_set.h [hunk 5]
```diff
@@ -91,6 +91,15 @@ class BasicLabelSet {
...
   }
 
+  template <class LabelSet>
+  PROMPP_ALWAYS_INLINE void append_sorted(const LabelSet& label_set) {
+    labels_.reserve(labels_.size() + label_set.size());
+
+    for (const auto& label : label_set) {
+      append(static_cast<std::string_view>(label.first), static_cast<std::string_view>(label.second));
+    }
+  }
```
Comment (vporoshok) [inline pp/primitives/label_set.h:103]: nit: Unlike `add()`, `append_sorted` does not sort or validate input order. If the caller ever passes an unsorted label set, the resulting `BasicLabelSet` will be silently corrupt.

Consider adding a debug-mode assertion, e.g.:
```cpp
assert(std::is_sorted(label_set.begin(), label_set.end(),
  [](const auto& a, const auto& b) { return a.first < b.first; }));
```
Or at minimum, document the precondition in a comment.

File: pp/wal/hashdex/go_head_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ protected:
+  using Lss = PromPP::Primitives::SnugComposites::LabelSet::EncodingBimap<BareBones::Vector>;
+
+  GoHead<Lss> hashdex_;
+  Lss lss_;
+  DataStorage data_storage_;
+  Encoder<> encoder_{data_storage_};
+};
+
+TEST_F(GoHeadFixture, Test) {
```
Comment (vporoshok) [inline pp/wal/hashdex/go_head_tests.cpp:35]: The C++ test covers the happy path well. However, there are no Go-level tests for the new path:
- `NewGoHeadHashdex` construction and finalization
- `AppendGoHeadHashdex` end-to-end through the adapter
- `BatchStorage.Commit` / `CommitWithState` via the new `GoHeadHashdex` path
- Edge case: commit when no data was appended (`samplesAdded == 0`)

Would be good to add at least a basic integration test on the Go side.

File: pp/entrypoint/wal_hashdex.cpp (hunk 2 above)
Comment (cherep58) [reply pp/entrypoint/wal_hashdex.cpp:214]: `prompp_wal_go_head_hashdex_presharding` doesn't thrown processable exceptions.
If an exception is thrown inside `prompp_wal_go_head_hashdex_presharding` process execution will be aborted

**Review by vporoshok — APPROVED** [review]
